scripts/sensor/_04b_time_frequency_csp.py

```diff
@@ -0,0 +1,436 @@
+                            check=False)
+
+    epochs = mne.read_epochs(fname_epochs)
+    if cfg.analyze_channels:
+        # We special-case the average reference here to work around a situation
+        # where e.g. `analyze_channels` might contain only a single channel:
+        # `concatenate_epochs` below will then fail when trying to create /
+        # apply the projection. We can avoid this by removing an existing
+        # average reference projection here, and applying the average reference
+        # directly – without going through a projector.
+        if 'eeg' in cfg.ch_types and cfg.eeg_reference == 'average':
+            epochs.set_eeg_reference('average')
+        else:
+            epochs.apply_proj()
+        epochs.pick(cfg.analyze_channels)
+
+    if cfg.time_frequency_subtract_evoked:
+        epochs.subtract_evoked()
+
+    # Perform rank reduction via PCA.
+    #
+    # Select the channel type with the smallest rank.
+    # Limit it to a maximum of 100.
+    ranks = mne.compute_rank(inst=epochs, rank='info')
+    ch_type_smallest_rank = min(ranks, key=ranks.get)
+    rank = min(
+        ranks[ch_type_smallest_rank],
+        100
+    )
+    del ch_type_smallest_rank, ranks
+
+    msg = f'Reducing data dimension via PCA; new rank: {rank}.'
+    logger.info(**gen_log_kwargs(msg, subject=subject, session=session))
+    pca = UnsupervisedSpatialFilter(
+        PCA(rank),
+        average=False
+    )
+
+    # Classifier
+    csp = CSP(
+        n_components=4,  # XXX revisit
+        reg=0.1,         # XXX revisit
+        rank='info',
+    )
+    clf = make_pipeline(
+        csp,
+        LinearDiscriminantAnalysis(),
+    )
+    cv = StratifiedKFold(
+        n_splits=cfg.decoding_n_splits,
+        shuffle=True,
+        random_state=cfg.random_state,
+    )
+
+    # Loop over frequencies (all time points lumped together)
+    freq_name_to_bins_map = dict()
+    for freq_range_name, freq_range_edges in cfg.decoding_csp_freqs.items():
+        freq_bins = list(zip(freq_range_edges[:-1], freq_range_edges[1:]))
+        freq_name_to_bins_map[freq_range_name] = freq_bins
+
+    freq_decoding_table = pd.DataFrame(
+        columns=['subject', 'cond_1', 'cond_2', 'f_min', 'f_max',
+                 'freq_range_name', 'mean_crossval_score', 'metric']
+    )
+
+    for freq_range_name, freq_bins in freq_name_to_bins_map.items():
+        for freq_bin in freq_bins:
+            f_min, f_max = freq_bin
+            one_row = pd.DataFrame(
+                {
+                    'subject': subject,
+                    'cond_1': condition1,
+                    'cond_2': condition2,
+                    'f_min': f_min,
+                    'f_max': f_max,
+                    'freq_range_name': freq_range_name,
+                    'mean_crossval_score': np.nan,
+                    'metric': cfg.decoding_metric,
+                },
+                index=[0]
+            )
+            freq_decoding_table = pd.concat(
```

Comment:
I would avoid calling concat to allow memory as each iteration and use first a list

Reply:
I've adjusted the logic to only call it once on a list of dicts